python_modules/dagster/dagster/_serdes/serdes.py

```diff
@@ -314,6 +313,79 @@ def __whitelist_for_serdes(klass: T_Type) -> T_Type:
     return __whitelist_for_serdes
 
 
+###################################################################################################
+# Serializers
+###################################################################################################
+
+
+class RecursiveDescentContext:
+    """values are packed/unpacked by recursing from the top."""
+
+    def __init__(self, descent_path: str):
+        self._descent_path = descent_path
+
+    def descend(self, path_suffix: str) -> "RecursiveDescentContext":
+        return RecursiveDescentContext(self._descent_path + path_suffix)
+
+    def get_error_context(self) -> str:
+        return f"\nDescent path: {self._descent_path}"
+
+
+class BottomUpUnpackContext:
+    """values are unpacked "bottom up" via json load object_hook."""
+
+    def __init__(self):
+        self._observed_unknown_serdes_values: Set[UnknownSerdesValue] = set()
+
+    def assert_no_unknown_values(self, obj: T) -> T:
+        # fast path out when none have been observed
+        if not self._observed_unknown_serdes_values:
+            return obj
+
+        if isinstance(obj, UnknownSerdesValue):
+            raise DeserializationError(
+                f"{obj.message}\nThis error can occur due to version skew, verify processes are"
+                " running expected versions."
+            )
+        elif isinstance(obj, (list, set, frozenset)):
+            for inner in obj:
+                self.assert_no_unknown_values(inner)
+        elif isinstance(obj, dict):
+            for k, v in obj.items():
+                self.assert_no_unknown_values(k)
+                self.assert_no_unknown_values(v)
+
+        return obj
+
+    def observe_unknown_value(self, val: "UnknownSerdesValue") -> "UnknownSerdesValue":
+        self._observed_unknown_serdes_values.add(val)
+        return val
+
+    def clear_ignored_unknown_values(self, obj: T) -> T:
+        if isinstance(obj, UnknownSerdesValue):
+            self._observed_unknown_serdes_values.discard(obj)
+        elif isinstance(obj, (list, set, frozenset)):
+            for inner in obj:
+                self.clear_ignored_unknown_values(inner)
+        elif isinstance(obj, dict):
+            for k, v in obj.items():
+                self.clear_ignored_unknown_values(k)
+                self.clear_ignored_unknown_values(v)
+
+        return obj
```

Comment:
This somewhat complex pattern of observe/clear/assert was the only way I could figure out to 
* catch things in nested structures 
* fire the `handle_unpack_error` at the right time
* be performant